File: product-service/src/main/java/com/shop/productservice/model/entity/Product.java
package com.shop.productservice.model.entity;

import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.FieldType;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

@Document(collection = "products")
@Data
public class Product {

    @Id
    private String id;
    @Indexed
    private String name;
    @Indexed
    @Field(targetType = FieldType.DECIMAL128)
    private BigDecimal price;
    private int amountLeft;
    private String description;
    private String imageUrl;
    private boolean available = true;

    /**
     * integer - amount of user who rated
     * double - rating of product
     */
    private Map<Integer, Double> ratings;
    private int orders;
    private List<Category> categories;
}

File: frontend-service/src/app/model/product.ts
export interface Product {
    id: string;
    name: string;
    price: number;
    description: string;
    quantity: number;
    image: string;
    rating: number;
    soldItems: number;
    categories: string[];
}

export interface ProductCart {
  productId: string;
  name: string;
  image: string;
  price: number;
  quantity: number;
  totalQuantity: number;
  rated: boolean;
}

export interface ProductDetails {
  id: string;
  name: string;
  available?: boolean;
  price: number;
  quantity: number;
  rating: number;
  image?: string;
  soldItems: number;
  categories: string[];
}

export interface ProductInfo {
  id: string;
  name: string;
  price: number;
  quantity: number;
}

export interface ProductsResponse {
  totalElements: number;
  categories: string[];
  products: ProductDetails[];
}

export interface NewProduct {
  name: string;
  price: number;
  description: string;
  quantity: number;
  categories: string[];
}

export interface CategoryNew {
  name: string;
}

